### ISSUE TITLE
Introduce CLI tool

### ISSUE THREAD

**Issue by trentmwillis** [qunitjs/qunit#1024]
Lately, we've been doing a lot of work on the reporter interface for QUnit and I think that once we start shipping a standard TAP (or other) reporter, it'll lower the barrier to adoption of QUnit for non-browser projects (specifically thinking about Node projects).

However, I think one thing that would complement that effort and make it trivial to use QUnit in those cases would be to introduce a command-line tool similar to what Mocha provides. In particular, this tool should make running QUnit tests in Node easy without having to provide a custom script.

There are some tools currently available (such as node-qunit or even mocha-qunit-ui), but it would be great to have one backed by the project that doesn't require additional overhead from the project maintainers.

As an aside: I would also very much like to see this happen so that we can unify the testing story in the Ember community, where most developers use QUnit in the browser for applications, but Mocha in node for build tools or other things.

**Comment by jzaefferer** [comment]
Based on previous conversations with @leobalter, I think he's very much on board with this. I like it in general as well. I'd like to make sure we get a full picture of everything involved though.

A CLI tool needs:

1. A CLI API to specify files to run, with some reasonable default (look up files in `test/*` or something)
2. Direct support for file watching and rerunning tests, or good third-party integration, like `nodemon`
3. Multiple reporters like tap, JUnit-style XML, whatever else there is

I like how `ember test --server` does 1. and 2., while the reporters are rather bad (at least in `ember-cli: 2.4.3`), especially the failure output is hard to parse.

Speaking of ember, would ember-cli switch from testem to a native QUnit CLI?

I'm sure I'm still missing a lot here...

**Comment by trentmwillis** [comment]
> I'd like to make sure we get a full picture of everything involved though.

Most definitely. I think what you've outlined above is a solid candidate for a minimum viable product and what most users would expect. I'll try to put together something a bit more formal that outlines our expected API and usage behavior/constraints.

> Would ember-cli switch from testem to a native QUnit CLI?

Doubtful as I imagine we won't want to support running/interfacing with browsers, something that Testem does well and is required for web apps. Rather, I imagine that you could set up QUnit as a "launcher" for Testem in the event that you have code that can be tested in Node ([see here for an example for Mocha](https://github.com/testem/testem#running-tests-in-node-and-custom-process-launchers)).

**Comment by jzaefferer** [comment]
> I'll try to put together something a bit more formal that outlines our expected API and usage behavior/constraints.

Sounds good. Maybe you could also do some research on existing modules that we can use?

Somewhat related, do you see this as a separate `qunit-cli` module? Or would it be bundled with QUnit (aka qunitjs on npm)?

**Comment by trentmwillis** [comment]
> Maybe you could also do some research on existing modules that we can use?

Will do.

> do you see this as a separate `qunit-cli` module? Or would it be bundled with QUnit?

I'd prefer it be bundled with QUnit. We could potentially maintain it as a separate package if needed, but I think I'd like to see a state where developers can write and run their tests easily with one install.

**Comment by leobalter** [comment]
I'm +1 to publish it bundled.

Ok to maintain as a separate package, but only if shows necessary.

**Comment by platinumazure** [comment]
I can't imagine why we would need it separate-- don't we just need to write a `cli.js` file and have the `bin` entry in package.json point to it? And then it could just pull in the core QUnit API.

**Comment by trentmwillis** [comment]
Planning on starting work on this this week. Below is a brief outline of the proposed initial API broken up over several PRs:

```
Usage: qunit [options] [files]

Files:

  One or more glob expressions to match for test files to run

Options:

  -h, --help                              output usage information
  -V, --version                           output the version number

  PR2:
  -f, --filter <string>                   filter tests using <string>

  PR3:
  -R, --reporter <name>                   specify the reporter to use
  --reporters                             display available reporters

  PR4:
  -w, --watch                             watch input files for changes
```

**Comment by leobalter** [comment]
I like the api and it's great to split it in multiple parts.

I have comments/suggestions in those extra parts:

I believe the reporter part is more interesting than filters, we have `only` the api accepts a glob of files, this should be easily handled.

again, for reporters: why not `-r` (lowercase)? And if the reporter is not provided, or a wrong name is given, list the available options. `--reporters` doesn't feel intuitive, IMHO.

**Comment by trentmwillis** [comment]
> I believe the reporter part is more interesting than filters

I agree, but adding the `filter` option should be super simple, since it is already supported by QUnit. We essentially just need to set `QUnit.config.filter` property with whatever argument is passed in.

> for reporters: why not -r (lowercase)? And if the reporter is not provided, or a wrong name is given, list the available options. --reporters doesn't feel intuitive, IMHO.

Primarily mimicking Mocha's API. That said, I agree, we can adjust it for better discoverability.

Will update the original comment.

**Comment by trentmwillis** [comment]
Okay, all of the PRs for the initial feature set are now up and ready for review! Thanks to everyone who has helped review so far, it is greatly appreciated and I am super excited to ship this thing 😄 .

- https://github.com/qunitjs/qunit/pull/1115 - Introduce QUnit CLI
- https://github.com/qunitjs/qunit/pull/1122 - Support filtering tests
- https://github.com/qunitjs/qunit/pull/1123 - Support file watching
- https://github.com/qunitjs/qunit/pull/1126 - Support custom reporters
- https://github.com/qunitjs/qunit/pull/1127 - Support seeds

**Comment by leobalter** [comment]
This came so fast and they are solved already! That's an amazing work! 

![giphy 1](https://cloud.githubusercontent.com/assets/301201/24536529/0181527e-15a9-11e7-8d88-0544ebc3e382.gif)
